core/modules/replica/HttpCatalogsModule.cc

```diff
@@ -0,0 +1,198 @@
+                                                   string const& taskName,
+                                                   unsigned int workerResponseTimeoutSec) {
+    return Ptr(new HttpCatalogsModule(controller, taskName, workerResponseTimeoutSec));
+}
+
+
+HttpCatalogsModule::HttpCatalogsModule(Controller::Ptr const& controller,
+                                       string const& taskName,
+                                       unsigned int workerResponseTimeoutSec)
+    :   HttpModule(controller,
+                   taskName,
+                   workerResponseTimeoutSec) {
+}
+
+
+void HttpCatalogsModule::executeImpl(qhttp::Request::Ptr const& req,
+                                     qhttp::Response::Ptr const& resp,
+                                     string const& subModuleName) {
+
+    debug(__func__);
+
+    util::Lock lock(_catalogsMtx, "HttpCatalogsModule::" + string(__func__));
+
+    // Check if a cached report can be used
+    if (not _catalogsReport.is_null()) {
+
+        // Send what's available so far before evaluating the age of the cache
+        // to see if it needs to be upgraded in the background.
+        sendData(resp, _catalogsReport);
+
+        uint64_t lastReportAgeMs = PerformanceUtils::now() - _catalogsReportTimeMs;
+        if (lastReportAgeMs < 60 * 60 * 1000) return;
+
+    } else {
+
+        // Send a dummy report for now, then upgrade the cache
+        bool const dummyReport = true;
+        json result;
+        for (auto&& database: controller()->serviceProvider()->config()->databases()) {
+            result["databases"][database] = _databaseStats(database, dummyReport);
+        }
+        sendData(resp, result);
+    }
+
+    // Otherwise, get the fresh snapshot of the replica distributions
+    json result;
+    result["databases"] = json::object();
+
+    for (auto&& database: controller()->serviceProvider()->config()->databases()) {
+        result["databases"][database] = _databaseStats(database);
+    }
+
+    // Update the cache
+    _catalogsReport = result;
+    _catalogsReportTimeMs = PerformanceUtils::now();
+
+    sendData(resp, _catalogsReport);
+}
+
+
+json HttpCatalogsModule::_databaseStats(string const& database, bool dummyReport) const {
+
+    auto const config = controller()->serviceProvider()->config();
+    auto const databaseServices = controller()->serviceProvider()->databaseServices();
+
+    vector<unsigned int> chunks;
+    if (not dummyReport) databaseServices->findDatabaseChunks(chunks, database);
+
+    vector<ReplicaInfo> replicas;
+    if (not dummyReport) databaseServices->findDatabaseReplicas(replicas, database);
+
+    json result;
+    result["chunks"]["unique"] = chunks.size();
+    result["chunks"]["with_replicas"] = replicas.size();
+
+    map<string, map<string, size_t>> stats;
+    for (auto&& table: config->databaseInfo(database).partitionedTables) {
+        stats[table]["data_unique_in_chunks_data"] = 0;
```

Comment:
it seems inefficient to not cache a ref to `stats[table]` and operate on that.

Reply:
Good point! Fixed.